### PR TITLE
Add additional lattice models

### PR DESCRIPTION
**Pull request by ddhawan11**
**Context:**
Adds additional lattice templates for the lattice class.

**Description of the Change:**
Adds lattice models: Lieb, Cubic, Body centered cubic and Face centered cubic.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-70987]

### CODE REVIEW

**Review by austingmhuang — COMMENTED** [review]
Functionally speaking it looks good! Make sure to remove print statements and add docstrings.

File: pennylane/spin/lattice.py [hunk 1]
```diff
@@ -317,6 +317,73 @@ def _kagome(n_cells, boundary_condition=False, neighbour_order=1):
...
+
+    return lattice_cubic
+
+
+def _bcc(n_cells, boundary_condition=False, neighbour_order=1):
+    r"""Generates a body centered cubic lattice"""
+    vectors = math.eye(3)
+    positions = [[0, 0, 0], [0.5, 0.5, 0.5]]
+
+    n_cells = n_cells[0:3]
```
[the remark this answers is not in the document]
Comment (austingmhuang) [reply pennylane/spin/lattice.py:357]: But I'm aware this has already been merged for the other internal functions so 🤷

[the remark this answers is not in the document]
Comment (ddhawan11) [reply pennylane/spin/lattice.py]: How about we raise an error, if anything other than 1 is passed in the rest of the dimensions?

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -50,6 +50,9 @@
   unique representation of the object.
   [(#6167)](https://github.com/PennyLaneAI/pennylane/pull/6167)
 
+* More predefined lattice shapes can now be generated.
```
Comment (soranjh) [inline doc/releases/changelog-dev.md:53]: Please be a bit more specific.

File: pennylane/spin/lattice.py [hunk 3]
```diff
@@ -220,136 +217,72 @@ def add_edge(self, edge_indices):
...
+        "kagome",
+        "lieb",
+        "cubic",
+        "bcc",
+        "fcc",
+    ]:
         raise ValueError(
             f"Lattice shape, '{lattice}' is not supported."
-            f"Please set lattice to: chain, square, rectangle, honeycomb, triangle, or kagome"
+            f"Please set lattice to: chain, square, rectangle, honeycomb, triangle, kagome, lieb,"
```
Comment (soranjh) [inline pennylane/spin/lattice.py:252]: We probably quotation marks here? ie., "chain" instead of chain?

File: pennylane/spin/lattice.py [hunk 4]
```diff
@@ -220,136 +217,72 @@ def add_edge(self, edge_indices):
...
 
     Returns:
         lattice object.
     """
 
     lattice_shape = lattice.strip().lower()
 
-    if lattice_shape not in ["chain", "square", "rectangle", "honeycomb", "triangle", "kagome"]:
+    if lattice_shape not in [
+        "chain",
```
Comment (soranjh) [inline pennylane/spin/lattice.py:239]: We should make sure that the name is case insensitive, e.g., user can put "chain", "Chain" etc. You can update this check to convert the input string to lowercase and then check if it exists in this list.

File: tests/spin/test_lattice.py [hunk 5]
```diff
@@ -525,6 +525,231 @@ def test_add_edge():
                 (8, 9, 0),
             ],
         ),
+        (
+            "LIEB",
```
Comment (soranjh) [inline tests/spin/test_lattice.py:529]: Do we just test the edges of the lattice? How about the correctness of other attributes?

**Review by soranjh — CHANGES_REQUESTED** [review]
Thanks @ddhawan11, looks good overall. We just need stronger tests for the correctness of the lattices.

File: tests/spin/test_lattice.py (hunk 5 above)
Comment (ddhawan11) [reply tests/spin/test_lattice.py:529]: We can test number of sites, or coordinates for lattices. I kept this suite limited as we already tested that when adding tests for the lattice class itself.

Comment (soranjh) [reply tests/spin/test_lattice.py:529]: So the correctness of the lattice object is tested for all lattice types? If yes, then the current test is enough.

Comment (ddhawan11) [reply tests/spin/test_lattice.py:528]: Added more tests to cover the templates which were missing.

File: pennylane/spin/lattice.py [hunk 6]
```diff
@@ -220,136 +217,72 @@ def add_edge(self, edge_indices):
...
-
-
-# TODO Check the efficiency of this function with a dictionary instead.
 def _generate_lattice(lattice, n_cells, boundary_condition=False, neighbour_order=1):
     r"""Generates the lattice object for a given shape and n_cells.
 
     Args:
-        lattice (str): Shape of the lattice. Input Values can be ``'chain'``, ``'square'``, ``'rectangle'``, ``'honeycomb'``, ``'triangle'``, or ``'kagome'``.
+        lattice (str): Shape of the lattice. Input Values can be ``'chain'``, ``'square'``, ``'rectangle'``,
+              ``'honeycomb'``, ``'triangle'``, ``'kagome'``, ``'lieb'``, ``'cubic'``, ``'bcc'``, or ``'fcc'``.
```
Comment (obliviateandsurrender) [inline pennylane/spin/lattice.py:225]: Where do we give more information regarding these lattice structures in the docs?

Comment (ddhawan11) [reply pennylane/spin/lattice.py:225]: I don't understand, do you mean we should add wiki links/papers to these?

Comment (soranjh) [reply pennylane/spin/lattice.py:225]: We can have a separate PR to improve the docs overall and also expose Lattice to users. @ddhawan11 Please keep note of such recommendation for that future PR.

File: tests/spin/test_lattice.py [hunk 7]
```diff
@@ -539,3 +763,180 @@ def test_shape_error():
     lattice = "Octagon"
     with pytest.raises(ValueError, match="Lattice shape, 'Octagon' is not supported."):
         _generate_lattice(lattice=lattice, n_cells=n_cells)
+
+
+@pytest.mark.parametrize(
+    ("shape", "n_cells", "expected_number"),
```
Comment (soranjh) [inline tests/spin/test_lattice.py:769]: Please rename to n_sites_expected or something similar.

File: tests/spin/test_lattice.py [hunk 8]
```diff
@@ -539,3 +763,180 @@ def test_shape_error():
...
+)
+def test_num_sites_lattice_templates(shape, n_cells, expected_number):
+    r"""Test that the correct number of lattice points are generated for the given attributes"""
+    lattice = _generate_lattice(lattice=shape, n_cells=n_cells)
+    assert lattice.n_sites == expected_number
+
+
+@pytest.mark.parametrize(
+    # expected_points here were calculated manually.
+    ("shape", "n_cells", "expected_points"),
```
Comment (soranjh) [inline tests/spin/test_lattice.py:786]: Do we have such tests for all lattices or just for 3D ones?

**Review by soranjh — APPROVED** [review]
Thanks @ddhawan11, left some minor comments and questions.

File: doc/releases/changelog-dev.md [hunk 9]
```diff
@@ -51,6 +51,9 @@
   unique representation of the object.
   [(#6167)](https://github.com/PennyLaneAI/pennylane/pull/6167)
 
+* Predefined lattice shapes, lieb, cubic, bcc, and fcc, can now be generated.
```
Comment (obliviateandsurrender) [inline doc/releases/changelog-dev.md:54]:
```suggestion
* Predefined lattice shapes such as `lieb`, `cubic`, `bcc`, and `fcc` can now be generated.
```

File: pennylane/spin/lattice.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/spin/lattice.py]: What purpose does having `n_cells` longer than the intended shape serve? If nothing, we should err on the side of raising an error/warning about skipping the rest of the values.

**Review by obliviateandsurrender — APPROVED** [review]
Looks, good. Just a few minor suggestions.

File: pennylane/spin/lattice.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ddhawan11) [reply pennylane/spin/lattice.py]: It was to provide flexibility to the user. I think we can raise a warning, or raise an error if anything in the higher dimension is other than 0 or 1. Please let me know if you have a preference for one over the other.

File: tests/spin/test_lattice.py (hunk 8 above)
Comment (ddhawan11) [reply tests/spin/test_lattice.py:786]: yes, we test the rest of the cases when we test the lattice class.

File: pennylane/spin/lattice.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/spin/lattice.py]: I prefer a `warning` as it doesn't bother the rest of the workflow.

File: tests/spin/test_spin_hamiltonian.py [hunk 10]
```diff
@@ -401,7 +401,7 @@ def test_heisenberg_hamiltonian(shape, n_cells, j, expected_ham):
         ),
         (
             "square",
-            [2, 2, 1],
+            [2, 2],
```
Comment (soranjh) [inline tests/spin/test_spin_hamiltonian.py:404]: [2, 2, 1] is equivalent to [2, 2]?

**Review by soranjh — APPROVED** [review]
Thanks @ddhawan11, looks good to me. Just left a final question.

File: tests/spin/test_spin_hamiltonian.py (hunk 10 above)
Comment (ddhawan11) [reply tests/spin/test_spin_hamiltonian.py:404]: This was when we were letting the users pass larger dimensions for n_cells, now when we added the error statement, it doesn't work the same way.

File: pennylane/spin/lattice.py [hunk 11]
```diff
@@ -220,136 +217,86 @@ def add_edge(self, edge_indices):
...
-    return lattice_kagome
-
-
-# TODO Check the efficiency of this function with a dictionary instead.
 def _generate_lattice(lattice, n_cells, boundary_condition=False, neighbour_order=1):
     r"""Generates the lattice object for a given shape and n_cells.
 
     Args:
-        lattice (str): Shape of the lattice. Input Values can be ``'chain'``, ``'square'``, ``'rectangle'``, ``'honeycomb'``, ``'triangle'``, or ``'kagome'``.
+        lattice (str): Shape of the lattice. Input Values can be ``'chain'``, ``'square'``, ``'rectangle'``,
```
Comment (austingmhuang) [inline pennylane/spin/lattice.py:224]:
```suggestion
        lattice (str): Shape of the lattice. Input values can be ``'chain'``, ``'square'``, ``'rectangle'``,
```

File: pennylane/spin/lattice.py [hunk 12]
```diff
@@ -220,136 +217,86 @@ def add_edge(self, edge_indices):
...
+        neighbour_order (int): Specifies the interaction level for neighbors within the lattice.
+                               Default is 1, indicating nearest neighbour.
 
     Returns:
         lattice object.
     """
 
     lattice_shape = lattice.strip().lower()
 
-    if lattice_shape not in ["chain", "square", "rectangle", "honeycomb", "triangle", "kagome"]:
```
Comment (austingmhuang) [inline pennylane/spin/lattice.py:336]: Not a blocker but to me it is a little strange to rename from lattice to lattice_shape. Perhaps it should just be called lattice_shape in the function signature from the beginning?

**Review by austingmhuang — APPROVED** [review]
Nice work @ddhawan11. Left two very nitpicky comments.

File: pennylane/spin/lattice.py (hunk 12 above)
Comment (soranjh) [reply pennylane/spin/lattice.py:336]: Nice suggestion, but we will not change the signature at this point. But @ddhawan11 please record this suggestion to be considered in future improvements.

**Review by soranjh — APPROVED** [review]

**Pull request merged**